File: 1 Algorithmic Toolbox/1 Introduction/2 Max pairwise product/C++/max_pairwise_product.cpp
#include <iostream>
#include <vector>

using std::vector;
using std::cin;
using std::cout;

long long MaxPairwiseProduct(const vector<int> &numbers) {
	int n = numbers.size();

	int max_ind1 = -1;
	for (int i = 0; i < n; i++)
		if ( (max_ind1 == -1) || (numbers[i] > numbers[max_ind1]) )
			max_ind1 = i;

	int max_ind2 = -1;
	for (int i = 0; i < n; i++)
		if ( (i != max_ind1) && ((max_ind2 == -1) || (numbers[i] > numbers[max_ind2])) )
			max_ind2 = i;

	return (long long)numbers[max_ind1] * numbers[max_ind2];
}

int main() {
	int n;
	cin >> n;
	vector<int> numbers(n);
	for (int i = 0; i < n; ++i) {
		cin >> numbers[i];
	}

	long long result = MaxPairwiseProduct(numbers);
	cout << result << "\n";
	return 0;
}

File: 1 Algorithmic Toolbox/Homeworks/2 Algorithmic Warm-up/2 Max pairwise product/Python/max_pairwise_product.py
# Uses python3
n = int(input())
a = [int(x) for x in input().split()]
assert(len(a) == n)

max_ind1 = -1
for i in range(n):
    if max_ind1 == -1 or a[i] > a[max_ind1]:
        max_ind1 = i

max_ind2 = -1
for i in range(n):
    if i != max_ind1 and (max_ind2 == -1 or a[i] > a[max_ind2]):
        max_ind2 = i

result = a[max_ind1] * a[max_ind2]

print(result)
